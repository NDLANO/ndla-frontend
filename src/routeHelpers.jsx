/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import config from './config';

export function toSearch() {
  return '/search';
}

const removeUrn = string => string.replace('urn:', '');

export function getUrnIdsFromProps(props) {
  const {
    match: { params },
  } = props;
  return {
    subjectId: params.subjectId ? `urn:${params.subjectId}` : undefined,
    topicId: params.topicId ? `urn:${params.topicId}` : undefined,
    resourceId: params.resourceId
      ? `urn:resource:${params.resourceId}`
      : undefined,
    articleId: params.articleId,
  };
}

export function toSubjects() {
  return `/subjects`;
}

export function toLearningPath(id) {
  return `${config.learningPathDomain}/learningpaths/${id}/first-step`;
}
export function toArticle(articleId, resource, subjectTopicPath, filters = '') {
  const filterParams = filters.length > 0 ? `?filters=${filters}` : '';
  if (subjectTopicPath) {
    return `${toSubjects()}${subjectTopicPath}/${removeUrn(
      resource.id,
    )}${filterParams}`;
  }
  if (resource) {
    return `${toSubjects()}${resource.path}/${filterParams}`;
  }
  return `/article/${articleId}${filterParams}`;
}

export function toSubject(subjectId) {
  return `${toSubjects()}/${removeUrn(subjectId)}`;
}

export function toTopic(subjectId, filters, ...topicIds) {
  const urnFreeSubjectId = removeUrn(subjectId);
  if (topicIds.length === 0) {
    return toSubject(urnFreeSubjectId);
  }
  const urnFreeTopicIds = topicIds.filter(id => !!id).map(removeUrn);
  const filterParam =
    filters && filters.length > 0 ? `?filters=${filters}` : '';
  const t = `${toSubjects()}/${urnFreeSubjectId}/${urnFreeTopicIds.join(
    '/',
  )}${filterParam}`;
  return t;
}

export const toTopicPartial = (
  subjectId,
  filters = '',
  ...topicIds
) => topicId => toTopic(subjectId, filters, ...topicIds, topicId);

export function toBreadcrumbItems(
  rootName,
  subject,
  topicPath = [],
  resource,
  filters = '',
) {
  const filterParam = filters.length > 0 ? `?filters=${filters}` : '';
  const topicLinks = topicPath.map(topic => ({
    to: toSubjects() + topic.path + filterParam,
    name: topic.name,
  }));

  const resourceLink = resource
    ? [{ to: toSubjects() + resource.path + filterParam, name: resource.name }]
    : [];

  return [
    { to: '/', name: rootName },
    { to: toSubjects() + subject.path + filterParam, name: subject.name },
    ...topicLinks,
    ...resourceLink,
  ];
}

export function toLinkProps(linkObject) {
  const isLearningpath =
    linkObject.contentUri &&
    linkObject.contentUri.startsWith('urn:learningpath') &&
    linkObject.meta;
  return {
    to: isLearningpath
      ? toLearningPath(linkObject.meta.id)
      : toSubjects() + linkObject.path,
    target: isLearningpath ? '_blank' : undefined,
    rel: isLearningpath ? 'noreferrer noopener' : undefined,
  };
}
