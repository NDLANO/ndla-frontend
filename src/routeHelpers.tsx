/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { matchPath, RouteComponentProps } from 'react-router-dom';
import {
  PROGRAMME_PAGE_PATH,
  PROGRAMME_PATH,
  SUBJECT_PAGE_PATH,
  TOPIC_PATH,
} from './constants';
import { getProgrammeBySlug } from './data/programmes';
import { getSubjectLongName } from './data/subjects';
import { LocaleType } from './interfaces';

export function toSearch(searchString?: string) {
  return `/search?${searchString || ''}`;
}

export const removeUrn = (str?: string) => (str ? str.replace('urn:', '') : '');

export function getUrnIdsFromProps(props: {
  ndlaFilm?: boolean;
  match: RouteComponentProps<{
    subjectId?: string;
    topicPath?: string;
    topicId?: string;
    resourceId?: string;
    articleId?: string;
  }>['match'];
}) {
  const {
    ndlaFilm,
    match: { params },
  } = props;
  const paramSubjectId = params.subjectId
    ? `urn:subject:${params.subjectId}`
    : undefined;
  const subjectId = ndlaFilm ? `urn:subject:20` : paramSubjectId;
  const topics = params.topicPath?.split('/') || [];
  const topicList = topics.map(t => `urn:${t}`);
  const topicId = params.topicId ? `urn:${params.topicId}` : undefined;
  if (topicId) {
    topicList.push(topicId);
  }

  return {
    subjectId,
    topicList,
    resourceId: params.resourceId
      ? `urn:resource:${params.resourceId}`
      : undefined,
    articleId: params.articleId,
    topicId: topicList[topicList.length - 1],
  };
}

function toLearningpaths() {
  return '/learningpaths';
}

type Resource = {
  path: string;
  id: string;
};

export function toLearningPath(
  pathId?: string,
  stepId?: string,
  resource?: Resource,
) {
  if (resource) {
    return stepId ? `${resource.path}/${stepId}` : resource.path;
  }
  if (pathId && stepId) {
    return `${toLearningpaths()}/${pathId}/steps/${stepId}`;
  }
  if (pathId) {
    return `${toLearningpaths()}/${pathId}`;
  }
  return toLearningpaths();
}

export function toArticle(
  articleId: number,
  resource: Resource,
  subjectTopicPath: string,
) {
  if (subjectTopicPath) {
    return `${subjectTopicPath}/${removeUrn(resource.id)}`;
  }
  if (resource) {
    return resource.path;
  }
  return `/article/${articleId}`;
}

export function toSubject(subjectId: string) {
  return `/${removeUrn(subjectId)}`;
}

export function toTopic(subjectId: string, ...topicIds: string[]) {
  const urnFreeSubjectId = removeUrn(subjectId);
  if (topicIds.length === 0) {
    return toSubject(urnFreeSubjectId);
  }
  const urnFreeTopicIds = topicIds.filter(id => !!id).map(removeUrn);
  const t = fixEndSlash(`/${urnFreeSubjectId}/${urnFreeTopicIds.join('/')}`);
  return t;
}

export const toTopicPartial = (subjectId: string, ...topicIds: string[]) => (
  topicId: string,
) => toTopic(subjectId, ...topicIds, topicId);

type Subject = {
  id?: string | undefined;
  name?: string | undefined;
  to?: string;
};

export function toBreadcrumbItems(
  rootName: string,
  paths: Subject[],
  locale: string = 'nb',
) {
  // henter longname fra filter og bruk i stedet for fÃ¸rste ledd i path
  const subject = paths[0];
  const longName: string = getSubjectLongName(subject?.id, locale);
  const breadcrumbSubject = {
    ...subject,
    name: longName || subject?.name,
  };

  const prelinks = [breadcrumbSubject, ...paths.splice(1)];

  const links = prelinks
    .filter(Boolean)
    .reduce(
      (links: Subject[], item) => [
        ...links,
        {
          to:
            (links.length ? links?.[links.length - 1]?.to : '') +
            '/' +
            removeUrn(item.id),
          name: item.name,
        },
      ],
      [],
    )
    .map(link => {
      // making sure we have the ending slash in breadCrumbs and it dosen't contain it allready
      if (link.to) {
        link.to = fixEndSlash(link.to);
      }
      return link;
    });
  return [{ to: '/', name: rootName }, ...links];
}

export function fixEndSlash(link: string) {
  const pattern = new RegExp(/resource/gi);
  if (link && !pattern.test(link) && !/\/$/.test(link)) {
    link = `${link}/`;
  }
  return link;
}

type LinkObject = {
  contentUri?: string;
  meta?: object;
  path?: string;
};

export function toLinkProps(linkObject: LinkObject) {
  const isLearningpath =
    linkObject.contentUri &&
    linkObject.contentUri.startsWith('urn:learningpath') &&
    linkObject.meta;
  return {
    to: isLearningpath ? toLearningPath() + linkObject.path : linkObject.path,
  };
}

export function toProgramme(programmePath: string) {
  return `${PROGRAMME_PATH}/${programmePath}`;
}

export function toProgrammeSubject(
  programmePath: string,
  subjectId: string,
  topicIds: string[],
) {
  return `${toProgramme(programmePath)}${toTopic(subjectId, ...topicIds)}`;
}

export function isSubjectPagePath(pathname: string) {
  const match = matchPath(pathname, SUBJECT_PAGE_PATH);
  if (match) {
    return match.isExact;
  }
  return false;
}

export function isTopicPath(pathname: string) {
  const match = matchPath(pathname, TOPIC_PATH);
  if (match) {
    return match.isExact;
  }
  return false;
}

export function getProgrammeByPath(pathname: string, locale: LocaleType) {
  const match = matchPath<{ programme?: string }>(
    pathname,
    PROGRAMME_PAGE_PATH,
  );
  if (match) {
    return getProgrammeBySlug(match.params.programme, locale);
  }
  return null;
}
