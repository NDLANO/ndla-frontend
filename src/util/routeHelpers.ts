/**
 * Copyright (C) 2021 -present, NDLA
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export function toLogin() {
  return '/login';
}

export function toLogout() {
  return '/logout';
}

export function toLogoutSession() {
  return '/logout/session';
}

export function toHome() {
  return '/';
}
