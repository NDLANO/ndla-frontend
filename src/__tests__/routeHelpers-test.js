/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { toBreadcrumbItems, isSubjectPagePath } from '../routeHelpers';

const subject = {
  id: 'urn:subject:9',
  name: 'Historie vg2 og vg3',
  path: '/subject:9',
};
const topics = [
  {
    id: 'urn:topic:1:179373',
    name: 'Om Utforskeren ',
    path: '/subject:9/topic:1:179373',
  },
  {
    id: 'urn:topic:1:170165',
    name: 'Samfunnsfaglige tenkemÃ¥ter',
    path: '/subject:3/topic:1:179373/topic:1:170165',
  },
];

const resource = {
  id: 'urn:resource:1:168389',
  name: 'Utforskeren',
  path: '/subject:9/topic:1:179373/topic:1:170165/resource:1:168389',
};

test('breadcrumb items from subject ', () => {
  expect(toBreadcrumbItems('Home', [subject])).toMatchSnapshot();
});

test('breadcrumb items from from subject and topicpath', () => {
  expect(toBreadcrumbItems('Home', [subject, ...topics])).toMatchSnapshot();
});

test('breadcrumb items from from subject, topicpath and resouce', () => {
  expect(
    toBreadcrumbItems('Home', [subject, ...topics, resource]),
  ).toMatchSnapshot();
});

test('is pathname a subject page path', () => {
  expect(isSubjectPagePath('/subjects/subject:1')).toBe(true);
  expect(isSubjectPagePath('/subjects/subject:134')).toBe(true);
  expect(isSubjectPagePath('/subjects/subject:1/')).toBe(true);
  expect(isSubjectPagePath('/subjects/subject:1/topic:1:186460')).toBe(true);
  expect(
    isSubjectPagePath('/subjects/subject:1/topic:1:184105/topic:1:184106'),
  ).toBe(true);
  expect(
    isSubjectPagePath(
      '/subjects/subject:1/topic:1:184105/topic:1:184106/topic:1:184107/resource:1:62382',
    ),
  ).toBe(true); // not exactly right, but must do for now.
});
