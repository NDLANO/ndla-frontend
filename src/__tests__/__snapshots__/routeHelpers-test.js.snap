// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breadcrumb items from from subject and topicpath 1`] = `
[
  {
    "name": "Home",
    "to": "/",
  },
  {
    "name": "Historie vg2 og vg3",
    "to": "/subject:9/",
  },
  {
    "name": "Om Utforskeren ",
    "to": "/subject:9/topic:1:179373/",
  },
  {
    "name": "Samfunnsfaglige tenkemåter",
    "to": "/subject:9/topic:1:179373/topic:1:170165/",
  },
]
`;

exports[`breadcrumb items from from subject, topicpath and resouce 1`] = `
[
  {
    "name": "Home",
    "to": "/",
  },
  {
    "name": "Historie vg2 og vg3",
    "to": "/subject:9/",
  },
  {
    "name": "Om Utforskeren ",
    "to": "/subject:9/topic:1:179373/",
  },
  {
    "name": "Samfunnsfaglige tenkemåter",
    "to": "/subject:9/topic:1:179373/topic:1:170165/",
  },
  {
    "name": "Utforskeren",
    "to": "/subject:9/topic:1:179373/topic:1:170165/resource:1:168389",
  },
]
`;

exports[`breadcrumb items from subject  1`] = `
[
  {
    "name": "Home",
    "to": "/",
  },
  {
    "name": "Historie vg2 og vg3",
    "to": "/subject:9/",
  },
]
`;
